### ISSUE TITLE
Paramspec issue in generic class with __init__

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3306]
**Describe the bug**
I have a small collection of examples of unexpected type errors/lack of type errors when using generic class with paramspec and P.args/P.kwargs on `__init__`.

```python
from typing import Callable, Generic, TypeVar

from typing_extensions import Concatenate, ParamSpec

P = ParamSpec("P")
T1 = TypeVar("T1")
T2 = TypeVar("T2")


def add_k(x: int, k: int) -> int:
    return x + k


class BadExecute1(Generic[P, T2]):
    def __init__(self, fn: Callable[P, T2], *args: P.args, **kwargs: P.kwargs) -> None:
        self.fn = fn
        self.args = args
        self.kwargs = kwargs

    def __call__(self) -> T2:
        return self.fn(*self.args, **self.kwargs)


bad_processor1 = BadExecute1(add_k)  # No type errors when there should be 1 for missing x/k.
print(bad_processor1())

bad_processor2 = BadExecute1(add_k, x=3, k=2)  # Two false positive type errors of no parameter x/k.
print(bad_processor2())


class BadExecute2(Generic[P, T1, T2]):
    def __init__(self, fn: Callable[Concatenate[T1, P], T2], *args: P.args, **kwargs: P.kwargs) -> None:
        self.fn = fn
        self.args = args
        self.kwargs = kwargs

    def __call__(self, value: T1) -> T2:
        return self.fn(value, *self.args, **self.kwargs)


bad_processor3 = BadExecute2(add_k)  # No type error even though k is missing.
print(bad_processor3(1))

bad_processor4 = BadExecute2(add_k, k=2)  # False positive error of no parameter named k.
print(bad_processor4(1))
```

Examples 1/3 will crash at runtime but have report no type errors. Examples 2/4 report type errors, but should be correct.

I'm guessing this is constructor specific as a similar example where P.args/P.kwargs are passed in a different function is handled well,

```python
# Type checks as expected
class GoodExecute1(Generic[P, T2]):
    def __init__(self, fn: Callable[P, T2]) -> None:
        self.fn = fn

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T2:
        return self.fn(*args, **kwargs)

good_processor1 = GoodExecute1(add_k)
print(good_processor1(1, 2))
```

**VS Code extension or command-line**
pyright cli 1.1.235

**Comment by erictraut** [comment]
How is `add_k` declared? Your example doesn't have a function declaration for this symbol.

**Comment by hmc-cs-mdrissi** [comment]
Sorry about that, I've edited it in along with imports/typevar definitions I accidentally forgot.

**Comment by erictraut** [comment]
Yes, I agree these are bugs. And you are correct that this is constructor-specific. Constructors contain a bunch of special-case logic because, unlike other methods, 1) they can bind class-scoped type variables and param specs and 2) the `__inti__` and `__new__` methods both need to be checked for consistency. This special-case handling is causing problems for param spec evaluation.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.236, which I just published. It will also be included in the next release of pylance.
